sdk/pinocchio/src/sysvars/slot_hashes/mod.rs

```diff
@@ -0,0 +1,348 @@
+            .binary_search_by(|probe_entry| probe_entry.slot().cmp(&target_slot).reverse())
+            .ok()
+    }
+
+    /// Returns a `&[SlotHashEntry]` view into the underlying data.
+    ///
+    /// The constructor (in the safe path that called `parse_and_validate_data`)
+    /// or caller (if unsafe `new_unchecked` path) is responsible for ensuring
+    /// the slice is big enough and properly aligned.
+    #[inline(always)]
+    fn as_entries_slice(&self) -> &[SlotHashEntry] {
+        let len = self.len();
+        debug_assert!(self.data.len() >= NUM_ENTRIES_SIZE + len * ENTRY_SIZE);
+
+        unsafe {
+            // SAFETY: The slice begins `NUM_ENTRIES_SIZE` bytes into `self.data`, which
+            // is guaranteed to have at least `len * ENTRY_SIZE` additional bytes. The
+            // pointer is properly aligned for `SlotHashEntry` because the struct is
+            // `repr(C)` and the original slice is aligned to at least u64.
+            // Therefore the constructed slice is valid
+            // for `len` elements for the lifetime of `&self`.
+            from_raw_parts(
+                self.data.as_ptr().add(NUM_ENTRIES_SIZE) as *const SlotHashEntry,
+                len,
+            )
+        }
+    }
+
+    /// Returns a reference to the entry at `index` **without** bounds checking.
+    ///
+    /// # Safety
+    /// Caller must guarantee that `index < self.len()`.
+    #[inline(always)]
+    pub unsafe fn get_entry_unchecked(&self, index: usize) -> &SlotHashEntry {
+        debug_assert!(index < self.len());
+        &self.as_entries_slice()[index]
+    }
+}
+
+impl<'a, T: Deref<Target = [u8]>> IntoIterator for &'a SlotHashes<T> {
+    type Item = &'a SlotHashEntry;
+    type IntoIter = core::slice::Iter<'a, SlotHashEntry>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.as_entries_slice().iter()
+    }
+}
+
+impl<'a> SlotHashes<Ref<'a, [u8]>> {
+    /// Creates a `SlotHashes` instance by safely borrowing data from an `AccountInfo`.
+    ///
+    /// This function verifies that:
+    /// - The account key matches the `SLOTHASHES_ID`
+    #[inline(always)]
+    pub fn from_account_info(account_info: &'a AccountInfo) -> Result<Self, ProgramError> {
+        if account_info.key() != &SLOTHASHES_ID {
+            return Err(ProgramError::Custom(ERR_WRONG_ACCOUNT_KEY));
+        }
+
+        let data_ref = account_info.try_borrow_data()?;
+
+        // SAFETY: The account was validated to be the `SlotHashes` sysvar.
+        Ok(unsafe { SlotHashes::new_unchecked(data_ref) })
+    }
+}
+
+#[cfg(feature = "std")]
+impl SlotHashes<Box<[u8]>> {
+    /// Allocates a buffer and fetches SlotHashes sysvar data via syscall.
+    ///
+    /// # Safety
+    /// The caller must ensure the buffer pointer is valid for MAX_SIZE bytes.
+    /// The syscall will write exactly MAX_SIZE bytes to the buffer.
+    #[inline(always)]
+    unsafe fn fetch_into_buffer(buffer_ptr: *mut u8) -> Result<(), ProgramError> {
```

Comment:
I think a `fetch` with this signature is probably more useful:
```rust
pub fn fetch() -> Result<Self, ProgramError>;
```
The `fetch_into_buffer` is more like an "advance" helper for people allocating their own buffer, which might not be that common.

Reply:
There was a `fetch_into` as well:D in raw.rs - with `buffer_ptr` as its first argument. I've renamed this one to `fill_from_sysvar` in eb86bd2